src/cache_cf.cc

```diff
@@ -1014,87 +1017,133 @@ parse_obsolete(const char *name)
     }
 
-    double d = xatof(token);
-
-    time_msec_t m = u; /* default to 'units' if none specified */
+    return true;
+}
 
-    if (d) {
-        if ((token = ConfigParser::PeekAtToken()) && (m = parseTimeUnits(token, allowMsec))) {
-            (void)ConfigParser::NextToken();
+/// Wraps parseSingleTimeUnit() and parses defaultUnitName if it failed to parse unitName.
+/// \returns true iff unitName was successfully parsed.
+template <class MinimalUnit>
+bool
+parseTimeUnits(const char *unitName, const char *defaultUnitName, std::chrono::nanoseconds &ns)
+{
+    if (unitName && parseSingleTimeUnit<MinimalUnit>(unitName, ns))
+        return true;
 
-        } else if (!expectMoreArguments) {
-            self_destruct();
-            return;
+    if (defaultUnitName) {
+        const auto defaultParsed = parseSingleTimeUnit<MinimalUnit>(defaultUnitName, ns);
+        assert(defaultParsed);
+        debugs(3, DBG_CRITICAL, "WARNING: No units on '" << config_input_line << "', assuming " << defaultUnitName);
+        return false;
+    }
 
-        } else {
-            token = NULL; // show default units if dying below
-            debugs(3, DBG_CRITICAL, "WARNING: No units on '" << config_input_line << "', assuming " << d << " " << units);
-        }
-    } else
-        token = NULL; // show default units if dying below.
+    if (unitName)
+        debugs(3, DBG_CRITICAL, "FATAL: unknown time unit '" << unitName << "'");
+    else
+        debugs(3, DBG_CRITICAL, "FATAL: missing time unit");
 
-    *tptr = static_cast<time_msec_t>(m * d);
+    self_destruct();
+    return false;
+}
 
-    if (static_cast<double>(*tptr) * 2 != m * d * 2) {
-        debugs(3, DBG_CRITICAL, "FATAL: Invalid value '" <<
-               d << " " << (token ? token : units) << ": integer overflow (time_msec_t).");
+static void
+CheckTimeOverflow(const double value, const std::chrono::nanoseconds &unit) {
+    const auto maxNanoseconds = std::chrono::nanoseconds::max().count();
+    if (value > maxNanoseconds/static_cast<double>(unit.count())) {
+        const auto maxYears = maxNanoseconds/(HoursPerYear*3600*1000000000);
+        debugs(3, DBG_CRITICAL, "FATAL: The time must be less than " << maxYears << " years");
         self_destruct();
     }
 }
 
-static uint64_t
-parseTimeUnits(const char *unit, bool allowMsec)
+/// Parses a time specification from the config file and
+/// returns the time as a chrono duration object of 'TimeUnit' type.
+/// \param defaultUnitName will be used if the time unit is missing,
+/// it should passed only if the configuration paramter being parsed
+/// excpects more arguments after time unit.
+template <class TimeUnit>
+TimeUnit
+parseTimeLine(const char *defaultUnitName = nullptr)
 {
-    if (allowMsec && !strncasecmp(unit, T_MILLISECOND_STR, strlen(T_MILLISECOND_STR)))
-        return 1;
-
-    if (!strncasecmp(unit, T_SECOND_STR, strlen(T_SECOND_STR)))
-        return 1000;
+    char *token = ConfigParser::NextToken();;
```

Comment:
```suggestion
    const auto token = ConfigParser::NextToken();
```
or
```suggestion
    auto token = ConfigParser::NextToken();;
```

However, since the function may use _two_ tokens, I suspect that we should not call one of them "token". Something like "valueToken" may work better.